### ISSUE TITLE
Pouchdb as a dependency

### ISSUE THREAD

**Issue by y-lohse** [cozy/cozy-client-js#88]
We are using pouch in the [offline module](https://github.com/cozy/cozy-client-js/blob/master/src/offline.js#L1), but it's declared as a [dev dep](https://github.com/cozy/cozy-client-js/blob/master/package.json#L53), so it's not always installed. I believe it should be declared as a regular dependency.

I'd be happy to supply a PR but I'm not sure how we should declare `pouchdb-find` and `pouchdb-adapter-memory`, as normal deps or as peer deps?

Ping @kosssi and @enguerran

**Comment by kosssi** [comment]
I'm ok to load pouchdb on dep of cozy-client-js.

In the last release of [pouchdb](https://github.com/pouchdb/pouchdb/releases/tag/6.1.2) pouchdb-find is include so is not necessary ;)

 pouchdb-adapter-memory is specific to test cozy-client-js offline.

**Comment by y-lohse** [comment]
Ok, thanks! Following our conversation, just pinging @m4dz to be sure: pouchdb isn't very big, but this change would still mean we'd bundle it along with cozy-client-js, as far as I'm aware. Is that ok?

I mean, if it's used in the module, I strongly believe it should be declared as a regular dependency, but let's get everyone's input on this.

**Comment by nono** [comment]
It is 38MB to download. I'd prefer to keep it as an optional dependency.

**Comment by y-lohse** [comment]
So there are 2 aspects here: 

1. if you use the offline part of cozy-client-js, you *need* pouchdb installed somewhere. IMO this shouldn't be the host package's job, but why not.
2. If an app imports cozy-client-js as part of it's build process but doesn't have pouchdb as a dependency, the build breaks. This is happening with photos at the moment. Adding pouchdb as a dependency to photos, despite not using it in any way, seems wrong. But maybe there's a way to fix the build without pouchdb?

**Comment by enguerran** [comment]
Are [peerDependencies](https://nodejs.org/en/blog/npm/peer-dependencies/#the-solution-peer-dependencies) not a good solution for such a library?

What is the wanted behavior?

- you don't want to use any kind of local persistence ⇒ no pouchDB is added, and the build do not break
- you want to use local persistence ⇒ you need to add pouchDB somewhere in your host bundle

The fact `cozy-photo-v3`'s build breaks if there is no pouchDB installed is not good and we may fix `cozy-client-js` to not require pouchDB mostly if we don't want local persistence.

**Comment by y-lohse** [comment]
I'm not very familiar with peerDependencies, but yes, it seems like this would be a good solution? The behavior you described is correct.

**Comment by nono** [comment]
IIRC, `peerDependencies` are not optional

**Comment by nono** [comment]
Maybe we can have all the code in this git repository, but publish two npm packages: one light, and one with offline support (and a dependency on pouchdb)

**Comment by m4dz** [comment]
My two cts :smile: :

- pouchdb & pouchdb-find as `peerDependencies` so they are installed along side the cozy-client-js in every apps
- remove [PouchDB* requires](https://github.com/cozy/cozy-client-js/blob/master/src/offline.js#L1-L2) to prevent Pouch included in each builds so keep the cozy-client-js lib lightweight
- update [`webpack.target.mobile.js`](https://github.com/cozy/cozy-files-v3/blob/master/config/webpack.target.mobile.js) target config to imports `PouchDB*` deps when needed:

```js
module.exports = {
  ...
  module: {
    loaders: [{
      test: /offline$/,
      loader: "imports-loader?PouchDB=pouchdb&pouchdbFind=pouchdb-find"
    }]
  }
}
```

so PouchDB* will be available in offline.js for mobile build only.

**Comment by y-lohse** [comment]
Does the solution outlined by @m4dz work for everyone? If not please raise your concerns now, otherwise I'll go ahead and implement it 🌪

**Comment by enguerran** [comment]
@kosssi what do you think about @m4dz proposal?

In my opinion, it is a good solution.

**Comment by y-lohse** [comment]
Sorry, I've been busy with other things and haven't had time to work on this.
FYI, if you remove the `import` statements, the tests break, so I think we also need to change the webpack config of cozy-client-js to import pouch during the tests, but I haven't been able to take a stab at it.

**Comment by enguerran** [comment]
I would like we fix this issue as it becomes blocking for other project... (referenced just above)

**Comment by gregorylegarec** [comment]
Hey guys, have a look at my PR, what do you think ?

**Comment by gregorylegarec** [comment]
Fixed by #131
